### PR TITLE
Remove comma from function definition with a single argument

### PR DESCRIPTION
**Pull request by topecongiro**
This PR removes comma from function definition if it only has a single argument and the argument fits into a single line. Partially solves #1441?
Also this PR changes the output of test in `fn-custom-7.rs`, assuming that those commas are unnecessary. If it is preferable to preserve them, I will change this PR accordingly.

### CODE REVIEW

**Comment by nrc** [comment]
Thanks for looking at this, it has been bothering me a bit lately - I briefly looked into it but couldn't find out what was wrong.

In general, I prefer to make sure that we generate the right text, rather than generating almost the right thing and then mutating it afterwards. (This generally leads to simpler code in the long run, and often seems to end up with a much nicer design as a side effect). I wonder if it is possible to do that here?

Could you explain what is going on to cause this bug please? I'd like to understand better what is happening here.

(The changes to fn-custom-7.rs seem fine, the old output looks bad).

**Comment by topecongiro** [comment]
Thank you for your feedback!

The bug is caused by [this part](https://github.com/rust-lang-nursery/rustfmt/blob/master/src/items.rs#L1888-L1890). We would like to use `SeperatorTactic::Horizontal` when there is only a single argument and the argument does not span multiple lines. In order to make this decision, we need to know whether rewrite of generics contsins newline or not.

For example, the following function from tests/fn-custom-3.rs with large generics and a single argument requires trailing comma after the argument:
```rust
fn bar<
    'a: 'bbbbbbbbbbbbbbbbbbbbbbbbbbb,
    TTTTTTTTTTTTT,
    UUUUUUUUUUUUUUUUUUUU: WWWWWWWWWWWWWWWWWWWWWWWW
>(
    a: Aaaaaaaaaaaaaaa,
) {
    bar();
}
```

I would like to suggest rewriting the relavant code to something like the following:
```rust
let tactic = if !generic_str_has_newline && arg_items.len() <= 1 {
    DefinitiveListTactic::Horizontal
} else {
    definitive_tactic(&arg_items,
                      context.config.fn_args_density.to_list_tactic(),
                      one_line_budget)
};
```

**Comment by topecongiro** [comment]
Follow up of the previous post which was inadequate and incorrect 😞 

   Whether to add a trailing comma or not is determined by `needs_trailing_separator` ([inside `write_list`](https://github.com/rust-lang-nursery/rustfmt/blob/master/src/lists.rs#L206)) from the values of `SeparatorTacit` and `DefinitiveListTactic`:
```rust
fn needs_trailing_separator(separator_tactic: SeparatorTactic,
                            list_tactic: DefinitiveListTactic)
                            -> bool {
    match separator_tactic {
        SeparatorTactic::Always => true,
        SeparatorTactic::Vertical => list_tactic == DefinitiveListTactic::Vertical,
        SeparatorTactic::Never => false,
    }
}
```
In order to avoid the bug, we need to use 

1.`SeparatorTacitc::Never` or
 2.`SeparatorTactic::Vertical` and the value other than `DefinitiveListTactic::Vertical`.

Inside `write_args` where the `write_list` is called, `SeparatorTactic` is determined based on the value of `fn_args_layout`. This causes the bug since it does not consider the number of argument:
   ```rust
    let (indent, trailing_comma, end_with_newline) = match context.config.fn_args_layout {
        IndentStyle::Block => {
            (indent.block_indent(context.config), SeparatorTactic::Vertical, true)
        }
        IndentStyle::Visual => (arg_indent, SeparatorTactic::Never, false),
    };
   ```
   We can change this to something like:
   ```rust
   let (indent, trailing_comma, end_with_newline) = match context.config.fn_args_layout {
       IndentStyle::Block if !generics_str_contains_newline && arg_items.len() <= 1 => {
           (indent.block_indent(context.config), SeparatorTactic::Never, true)
       }
       IndentStyle::Block => {
           (indent.block_indent(context.config), SeparatorTactic::Vertical, true)
       }
       IndentStyle::Visual => (arg_indent, SeparatorTactic::Never, false),
   };
   ```
I am currently busy on [Shinsotsu-ikkatsu-saiyō](https://en.wikipedia.org/wiki/Simultaneous_recruiting_of_new_graduates), but will work on this in my spare time.

**Comment by topecongiro** [comment]
Updated with tests and a small change to close #1501.

File: tests/source/configs-fn_call_style-block.rs [hunk 1]
```diff
@@ -1,6 +1,9 @@
 // rustfmt-fn_call_style: Block
+// rustfmt-max_width: 120
```
Comment (nrc) [inline tests/source/configs-fn_call_style-block.rs]: I would prefer not to change max-width in this test, instead we ought to be able to change the code to work at the default width

**Comment by topecongiro** [comment]
Hi, thank you for you feedback! I updated the test.

**Comment by nrc** [comment]
Unfortunately the test seems to fail now?

**Comment by topecongiro** [comment]
Running `cargo test` locally finishes without an error. Looks like travis failed to install nightly, and tests for stable are failing in AppVeyor due to 'static_lifetime_in_statics' things.

**Comment by nrc** [comment]
I restarted Travis, lets see...

**Comment by topecongiro** [comment]
It's failing again...

**Comment by nrc** [comment]
Apparently there were nightly issues, I restarted Travis again (I'm sure this will be fine, but just to be safe...)

**Comment by nrc** [comment]
Yay, this time it worked :-)

**Pull request merged**